runtime/common/src/claims.rs

```diff
@@ -254,6 +280,80 @@ decl_module! {
 				<Vesting<T>>::insert(who, vs);
 			}
 		}
+
+		/// Make a claim to collect your DOTs.
+		///
+		/// The dispatch origin for this call must be _None_.
+		///
+		/// Unsigned Validation:
+		/// A call to claim is deemed valid if the signature provided matches
+		/// the expected signed message of:
+		///
+		/// > Ethereum Signed Message:
+		/// > (configured prefix string)(address)(statement)
+		///
+		/// and `address` matches the `dest` account; the `statement` must match that which is
+		/// expected according to your purchase arrangement.
+		///
+		/// Parameters:
+		/// - `dest`: The destination account to payout the claim.
+		/// - `ethereum_signature`: The signature of an ethereum signed message
+		///    matching the format described above.
+		/// - `statement`: The identity of the statement which is being attested to in the signature.
+		///
+		/// <weight>
+		/// The weight of this call is invariant over the input parameters.
+		/// - One `eth_recover` operation which involves a keccak hash and a
+		///   ecdsa recover.
+		/// - Four storage reads to check if a claim exists for the user, to
+		///   get the current pot size, to see if there exists a vesting schedule, to get the
+		///   required statement.
+		/// - Up to one storage write for adding a new vesting schedule.
+		/// - One `deposit_creating` Currency call.
+		/// - One storage write to update the total.
+		/// - Two storage removals for vesting and claims information.
+		/// - One deposit event.
+		///
+		/// Total Complexity: O(1)
+		/// ----------------------------
+		/// Base Weight: 622.6 µs
+		/// DB Weight:
+		/// - Read: Claims, Total, Claims Vesting, Vesting Vesting, Balance Lock, Account
+		/// - Write: Vesting Vesting, Account, Balance Lock, Total, Claim, Claims Vesting
+		/// </weight>
+		#[weight = T::DbWeight::get().reads_writes(7, 6) + 650_000_000]
+		fn claim_attest(origin,
+			dest: T::AccountId,
+			ethereum_signature: EcdsaSignature,
+			statement: StatementKind,
+		) {
+			ensure_none(origin)?;
+
+			let data = dest.using_encoded(to_ascii_hex);
+			let signer = Self::eth_recover(&ethereum_signature, &data, statement.to_text())
+				.ok_or(Error::<T>::InvalidEthereumSignature)?;
+			if let Some(s) = Signing::get(signer) {
+				ensure!(s == statement, Error::<T>::InvalidStatement);
+			}
+			Self::process_claim(signer, dest)?;
+		}
+
+		/// Attest to a statement, needed to finalize the claims process.
+		///
+		/// WARNING: Insecure unless your chain includes `PrevalidateAttests` as a `SignedExtension`.
+		#[weight = (
+			T::DbWeight::get().reads_writes(6, 6) + 650_000_000,
+			DispatchClass::Normal,
+			Pays::No
+		)]
+		fn attest(origin, statement: Vec<u8>) {
```

Comment:
Why don't we use here the `StatementKind`?

Reply:
i think legally speaking it would be much weaker, since the data being signed is only machine readable and entirely dependent on the current runtime.

Reply:
The idea of passing in a `Vec<u8>` is that there is notionally human readable content which is cryptographically signed.

Reply:
Why do we allow `StatementKind` in claim_attest then? Would make more sense to have both kind of attest sign the same thing.

Reply:
Yeah, that was the reason I asked for this.

Reply:
yeah, that should probably be switched over.

Reply:
I switched it